source/iot/MqttRequestResponseClient.cpp

```diff
@@ -0,0 +1,509 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/iot/MqttRequestResponseClient.h>
+
+#include <aws/crt/mqtt/Mqtt5Client.h>
+#include <aws/crt/mqtt/MqttConnection.h>
+
+#include <aws/common/ref_count.h>
+#include <aws/common/rw_lock.h>
+
+namespace Aws
+{
+    namespace Iot
+    {
+        namespace RequestResponse
+        {
+
+            class StreamReadLock
+            {
+              public:
+                StreamReadLock(struct aws_rw_lock *lock, struct aws_event_loop *protocolLoop)
+                    : m_lock(lock), m_taken(false)
+                {
+                    if (!aws_event_loop_thread_is_callers_thread(protocolLoop))
+                    {
+                        m_taken = true;
+                        aws_rw_lock_rlock(lock);
+                    }
```

Comment:
I think this `if` logic might be flawed.
If `StreamReadLock` is created by the event loop's thread, it won't lock `lock`. So, some writer from another thread will be able to obtain `lock` and get access to protected data while event loop thread is reading it.

Reply:
Aw crud